packages/babel-register/src/babel-core/track-deps.js

```diff
@@ -0,0 +1,94 @@
+"use strict";
+
+const { addHook } = require("pirates");
+
+const instances = new Map();
+
+let counter = 0;
+
+module.exports = exports = load;
+exports.wrap = wrap;
+
+/**
+ * This function exposes a way to load a module and inject logic to track
+ * what dependencies are loaded from there. This includes files that don't
+ * load right away, so the 'files' list should be treated as live data.
+ * If something inside the module triggers a call to `require()`
+ * at a later time, those items will also be included in 'files'.
+ */
+function load(
+  // The require() function to call when loading 'id'.
+  req: string => mixed,
+  // The name of the file that 'id' will be loaded relative to.
+  filename: string,
+  // The module to load.
+  id: string,
+  // A callback to call when resolving a module. If an object is returned,
+  // those exports will be used and that module and anything it loads will not
+  // be tracked in the 'files' Set.
+  overrideCallback: (
+    // The id of a module being loaded.
+    id: string,
+    // The name of the file that is loading the id.
+    filename: string,
+    // 'true' if the 'filename' is known to be in the same package that
+    // was originally passed to 'load'. 'true' as long as the file was
+    // reached using only relative paths within the original module.
+    inRoot: boolean,
+  ) => { exports: mixed } | void,
+) {
+  const files = new Set();
+
+  const uid = counter++;
+
+  instances.set(uid, {
+    files,
+    overrideCallback: overrideCallback,
+  });
+
+  return {
+    exports: wrap(req, filename, uid, true)(id),
+    files,
+  };
+}
+
+function wrap(req, filename, uid, inRoot) {
+  function babelRegisterRequire(id) {
+    const { files, overrideCallback } = instances.get(uid);
+
+    const override = overrideCallback(id, filename, inRoot);
+    if (override) return override.exports;
+
+    const depFilename = req.resolve(id);
+
+    files.add(depFilename);
+
+    const removeHook = addHook(
+      (code, loadingFilename) => {
+        return loadingFilename === depFilename
+          ? injectWrap(code, uid, inRoot && /^\.\.?[/\\]/.test(id))
+          : code;
+      },
+      { ignoreNodeModules: false },
+    );
+    try {
+      return req(id);
+    } finally {
+      removeHook();
+    }
+  }
+
+  // Ensure that require.resolve and other properties are carried over.
+  return Object.assign(babelRegisterRequire, req);
+}
+
+// For any file load kicked off by @babel/core or its dependencies has
+// a snippet injected into it to wrap the 'require' hook so that we
+// can avoid Babel triggering cyclical dependencies.
+function injectWrap(code, uid, inRoot) {
+  const override = `require(${JSON.stringify(
+    __filename,
+  )}).wrap(require, __filename, ${uid}, ${inRoot})`;
+
+  return `return (function(require){${code}\n}).call(this, ${override});`;
```

Comment:
top level return is icky 😅

Reply:
Yeah for some reason I was thinking it was needed, but it isn't is it.

Reply:
wonder how this affects the error messages (although that might just be messed up already)

Reply:
They'll be an extra scope in the debugger if you pause, but otherwise I don't think it will affect anything.